Runtime/Triangulator.cs

```diff
@@ -2698,60 +2684,144 @@ private static void Eigen(double2x2 matrix, out double2 eigval, out double2x2 ei
+            foreach (var p in positions)
+            {
+                min = math.min(p, min);
+                max = math.max(p, max);
+                com += p;
+            }
+
+            com /= positions.Length;
+            // Note: For integer coordinates we only translate points to the origin, but do not scale them.
+            return Translate(-com);
+        }
+    }
+
+    /// <summary>
+    /// Helper interface for math operations that are generic over the coordinate type.
+    /// </summary>
+    /// <typeparam name="T">The raw coordinate type for a single axis. For example float or int.</typeparam>
+    /// <typeparam name="T2">The 2D coordinate composed of Ts. For example float2.</typeparam>
+    /// <typeparam name="TFloat">A floating point value. Used for angles and fractional values. For example float.</typeparam>
+    /// <typeparam name="TBig">A squared coordinate. Used for squared distances and other products. For example float or long.</typeparam>
+    internal interface IUtils<T, T2, TFloat, TBig> where T : unmanaged where T2 : unmanaged where TFloat : unmanaged where TBig : unmanaged
+    {
+        /// <summary>
+        /// Converts a value to type T.
+        ///
+        /// Warning: This may cause loss of precision for very large integers.
+        /// </summary>
+        T Const(int v);
+        /// <summary>
+        /// Converts a value to type TBig.
+        ///
+        /// Warning: This may cause loss of precision, and is only safe for (not huge) integer constants.
+        /// </summary>
+        TBig ConstDistanceSq(float v);
+        /// <summary>
+        /// Converts a value to type TFloat.
+        ///
+        /// Warning: This may cause loss of precision, but it always supports fractional values.
+        /// </summary>
+        TFloat ConstFloat(float v);
         T EPSILON();
+        TBig EPSILON_SQ();
         T MaxValue();
         T2 MaxValue2();
+        TBig MaxDistSq();
         T2 MinValue2();
         T2 NewT2(T x, T y);
         T X(T2 v);
         T Y(T2 v);
         T Zero();
+        TBig ZeroSq();
+        bool SupportsMeshRefinement { get; }
 #pragma warning disable IDE1006
         T abs(T v);
+        TBig abs(TBig v);
         T add(T a, T b);
-        T alpha(T D, T dSquare, bool initial);
-        bool anyabslessthan(T a, T b, T c, T v);
-        T atan2(T v, T w);
+        TBig add(TBig a, TBig b);
+        TFloat alpha(TFloat D, TBig dSquare, bool initial);
+        bool anyabslessthan(TFloat a, TFloat b, TFloat c, TFloat v);
+        // Note: Takes TBig arguments, even if that's not technically correct for an atan2 function.
+        // It is used like atan2(a*C, b*C) where C is a length of type T.
+        TFloat atan2(TBig v, TBig w);
         T2 avg(T2 a, T2 b);
         T diff(T a, T b);
+        TBig diff(TBig a, TBig b);
         T2 diff(T2 a, T2 b);
-        T distancesq(T2 a, T2 b);
+        TBig distancesq(T2 a, T2 b);
```

Comment:
Based on previous comments, it seems to me that it's best to have uniform parameter types in all math methods, i.e. `T distancesq(T, T)` and seperate `TBig distancesq(TBig, TBig)` with additional casts `T -> TBig`. This should be more consistent but it can be cleanuped later in a separate PRs.

Reply:
But `distancesq(TBig, TBig)` does not make sense. That would be `distancesq(TBig2, TBig2)` which would require us to add another generic type.

Reply:
Ooopps, you are right, my mistake, I simplified it too much in my head!